packages/net/tcpconnection.pony

```diff
@@ -377,29 +417,36 @@ actor TCPConnection
 
         while _readable and not _shutdown_peer do
           // Read as much data as possible.
-          let len = @pony_os_recv[USize](_event, _read_buf.cstring(),
-            _read_buf.space()) ?
-
-          var next = _read_buf.space()
+          let len = @pony_os_recv[USize](
+            _event,
+            _read_buf.cstring().usize() + _read_len,
+            _read_buf.space() - _read_len) ?
 
           match len
           | 0 =>
             // Would block, try again later.
             _readable = false
             return
-          | _read_buf.space() =>
+          | _next_size =>
             // Increase the read buffer size.
-            next = _max_size.min(next * 2)
+            _next_size = _max_size.min(_next_size * 2)
           end
 
-          let data = _read_buf = recover Array[U8].undefined(next) end
-          data.truncate(len)
-          _notify.received(this, consume data)
+          _read_len = _read_len + len
+
+          if _read_len >= _expect then
+            let data = _read_buf = recover Array[U8] end
+            data.truncate(_read_len)
```

Comment:
@sylvanc - ah, I missed the fact that this is truncating on `_read_len`, not on `_expect`.

My bad.